File: packages/earthsci-toolkit/src/flatten.test.ts
import { describe, it, expect } from 'vitest'
import { flatten } from './flatten.js'
import type { EsmFile } from './types.js'

describe('flatten', () => {
  it('namespaces variables from a single model', () => {
    const file = {
      esm: '0.1.0',
      metadata: { name: 'test' },
      models: {
        Atmos: {
          variables: {
            T: { type: 'state' },
            k: { type: 'parameter' },
          },
          equations: [
            {
              lhs: { op: 'D', args: ['T'], wrt: 't' },
              rhs: { op: '*', args: ['k', 'T'] },
            },
          ],
        },
      },
    } as unknown as EsmFile

    const flat = flatten(file)
    expect(flat.stateVariables).toEqual(['Atmos.T'])
    expect(flat.parameters).toEqual(['Atmos.k'])
    expect(flat.metadata.sourceSystems).toEqual(['Atmos'])
    expect(flat.equations).toHaveLength(1)
    expect(flat.equations[0]!.sourceSystem).toBe('Atmos')
    expect(flat.equations[0]!.lhs).toContain('Atmos.T')
    expect(flat.equations[0]!.rhs).toContain('Atmos.k')
    expect(flat.equations[0]!.rhs).toContain('Atmos.T')
  })

  it('namespaces species and parameters from a reaction system', () => {
    const file = {
      esm: '0.1.0',
      metadata: { name: 'test' },
      reaction_systems: {
        Chem: {
          species: { O3: { units: 'mol/L' } },
          parameters: { k1: { units: '1/s' } },
          reactions: [
            {
              substrates: [{ species: 'O3', stoichiometry: 1 }],
              products: [],
              rate: { op: '*', args: ['k1', 'O3'] },
            },
          ],
        },
      },
    } as unknown as EsmFile

    const flat = flatten(file)
    expect(flat.stateVariables).toContain('Chem.O3')
    expect(flat.parameters).toContain('Chem.k1')
    expect(flat.metadata.sourceSystems).toEqual(['Chem'])
    expect(flat.equations.length).toBeGreaterThan(0)
  })

  it('records coupling rules in metadata', () => {
    const file = {
      esm: '0.1.0',
      metadata: { name: 'test' },
      models: {
        A: { variables: { x: { type: 'state' } }, equations: [] },
        B: { variables: { y: { type: 'parameter' } }, equations: [] },
      },
      coupling: [
        {
          type: 'variable_map',
          from: 'A.x',
          to: 'B.y',
          transform: 'identity',
        },
      ],
    } as unknown as EsmFile

    const flat = flatten(file)
    expect(flat.metadata.couplingRules).toHaveLength(1)
    expect(flat.metadata.couplingRules[0]).toContain('variable_map')
    expect(flat.variables['B.y']).toBe('A.x')
  })

  it('produces nested dot-namespacing for subsystems', () => {
    const file = {
      esm: '0.1.0',
      metadata: { name: 'test' },
      models: {
        Outer: {
          variables: { y: { type: 'state' } },
          equations: [],
          subsystems: {
            Inner: {
              variables: { x: { type: 'state' } },
              equations: [],
            },
          },
        },
      },
    } as unknown as EsmFile

    const flat = flatten(file)
    expect(flat.stateVariables).toContain('Outer.y')
    expect(flat.stateVariables).toContain('Outer.Inner.x')
  })
})

File: packages/esm-format-go/pkg/esm/flatten_test.go
package esm

import (
	"strings"
	"testing"
)

func TestFlatten_SingleModelNamespacesVariables(t *testing.T) {
	file := &EsmFile{
		Models: map[string]Model{
			"Atmos": {
				Variables: map[string]ModelVariable{
					"T": {Type: "state"},
					"k": {Type: "parameter"},
				},
				Equations: []Equation{},
			},
		},
	}

	flat, err := Flatten(file)
	if err != nil {
		t.Fatalf("Flatten: %v", err)
	}

	if !contains(flat.StateVariables, "Atmos.T") {
		t.Errorf("expected Atmos.T in state variables, got %v", flat.StateVariables)
	}
	if !contains(flat.Parameters, "Atmos.k") {
		t.Errorf("expected Atmos.k in parameters, got %v", flat.Parameters)
	}
	if !contains(flat.Metadata.SourceSystems, "Atmos") {
		t.Errorf("expected Atmos in source systems, got %v", flat.Metadata.SourceSystems)
	}
}

func TestFlatten_ReactionSystemNamespacesSpecies(t *testing.T) {
	file := &EsmFile{
		ReactionSystems: map[string]ReactionSystem{
			"Chem": {
				Species: map[string]Species{
					"O3": {},
				},
				Parameters: map[string]Parameter{
					"k1": {},
				},
				Reactions: []Reaction{},
			},
		},
	}

	flat, err := Flatten(file)
	if err != nil {
		t.Fatalf("Flatten: %v", err)
	}

	if !contains(flat.StateVariables, "Chem.O3") {
		t.Errorf("expected Chem.O3 in state variables, got %v", flat.StateVariables)
	}
	if !contains(flat.Parameters, "Chem.k1") {
		t.Errorf("expected Chem.k1 in parameters, got %v", flat.Parameters)
	}
}

func TestFlatten_RecordsCouplingRules(t *testing.T) {
	file := &EsmFile{
		Models: map[string]Model{
			"A": {
				Variables: map[string]ModelVariable{"x": {Type: "state"}},
				Equations: []Equation{},
			},
			"B": {
				Variables: map[string]ModelVariable{"y": {Type: "parameter"}},
				Equations: []Equation{},
			},
		},
		Coupling: []interface{}{
			VariableMapCoupling{
				Type:      "variable_map",
				From:      "A.x",
				To:        "B.y",
				Transform: "identity",
			},
		},
	}

	flat, err := Flatten(file)
	if err != nil {
		t.Fatalf("Flatten: %v", err)
	}

	if len(flat.Metadata.CouplingRules) == 0 {
		t.Fatalf("expected coupling rules to be recorded")
	}
	found := false
	for _, rule := range flat.Metadata.CouplingRules {
		if strings.Contains(rule, "variable_map") || strings.Contains(rule, "VariableMap") {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("expected variable_map rule, got %v", flat.Metadata.CouplingRules)
	}
}

func contains(haystack []string, needle string) bool {
	for _, h := range haystack {
		if h == needle {
			return true
		}
	}
	return false
}
